tokio/src/runtime/task/list.rs

```diff
@@ -140,34 +170,39 @@ impl<S: 'static> OwnedTasks<S> {
 
     /// Shuts down all tasks in the collection. This call also closes the
     /// collection, preventing new items from being added.
-    pub(crate) fn close_and_shutdown_all(&self)
+    /// The parameter start should be random among different worker threads
+    /// to reduce lock conflicts during shutdown.
+    /// Initiate shutting down the segment indexed by the start, and reset to 0
+    /// once the segment_size is reached, continuing until start - 1, it works like a ring.
+    pub(crate) fn close_and_shutdown_all(&self, start: usize)
     where
         S: Schedule,
     {
-        // The first iteration of the loop was unrolled so it can set the
-        // closed bool.
-        let first_task = {
-            let mut lock = self.inner.lock();
-            lock.closed = true;
-            lock.list.pop_back()
-        };
-        match first_task {
-            Some(task) => task.shutdown(),
-            None => return,
+        self.closing.store(true, Ordering::Release);
+        for i in start..self.get_segment_size() + start {
+            loop {
+                let mut lock = self.segment_inner(i);
+                match lock.pop_back() {
+                    Some(task) => {
+                        drop(lock);
+                        self.count.fetch_sub(1, Ordering::Relaxed);
+                        task.shutdown();
+                    }
+                    None => break,
+                };
+            }
         }
+        // we have shut down all tasks
+        self.shutdown.store(true, Ordering::Release)
```

Comment:
What is the purpose of this flag?

Reply:
I use `count: AtomicUsize` to represent the number of tasks, and wo do operation by `Ordering::Relaxed` everywhere for the best performance, but it has no `happens-before` relationship.

The shutdown flag is write by `Ordering::Release` and read by `Ordering::Acquire`, it can ensure a `happens-before` relationship.

I use this flag here to make sure the following assert can always pass:

```
debug_assert!(self.shared.owned.is_shutdown()); // do has a `happens-before` relationship.
debug_assert!(self.shared.owned.is_empty()); // has no `happens-before` relationship.
```

Reply:
Because in `multi_thread` mod, the lock is always obtained first, and then the closed flag of OwnedTasks is read. So 
 I would like to delete the shutdown flag.